writing/2020-08-08-Revisiting-application-structure.md

```diff
@@ -0,0 +1,294 @@
+   it's a reader where the behavior is defined by the object we feed in,
+   gotcha!).
+
+## Why readers became a burden
+There are a few drawbacks to this pattern. Let's start with the most glaring issues:
+
+- Granular control over interfaces becomes tedious due to the extra indirection
+  with the passed data
+
+- Certain things are difficult to implement, e.g:
+
+  ```haskell
+  class Foo m where
+    withCallback :: (a -> m b) -> m b
+  ```
+
+  This requires a _lot_ of lifting back and forth especially when the concrete
+  implementation is in `IO` and your interfaces are all in `m`
+
+- Lastly, and most important: it didn't turn out to be so easy to grok as we
+  thought
+
+## Looking at alternatives
+There are a couple of alternatives to this approach to dealing with effects. If
+you _want_ effect tracking in your types - there are a number of libraries that
+deal with this:
+
+- [polysemy](https://hackage.haskell.org/package/polysemy)
+- [fused-effects](https://hackage.haskell.org/package/fused-effects)
+
+Both of these are promising, but we're not really comfortable with the
+drawbacks to either one at the moment. In a nutshell - they're both great
+libraries, however, they're pretty advanced.
+
+Our old solution combined two things - and this was its main mistake. Either we
+should've said no to our interfaces and gone with something like the handler
+pattern - or we should've leaned fully into MTL.
+
+The drawback with handler pattern would that we can't be polymporphic, we
+really like for testing. The drawback with MTL is the `n^2` instances problem.
+
+Urgh! From our wishes on polymorphism it's clear we can't use the handler
+pattern. But can we use MTL if we solve the `n^2` issue? And what is the `n^2`
+issue?
+
+## The `n^2` issue
+When using monad transformers, you need to write the monad instances for all
+the different types of transformers. Here's an example from the [MTL source
+code][mtl-boilerplate]:
+
+```haskell
+instance MonadState s m => MonadState s (ExceptT e m) where
+  get = lift get
+  put = lift . put
+  state = lift . state
+
+instance MonadState s m => MonadState s (IdentityT m) where
+  get = lift get
+  put = lift . put
+  state = lift . state
+
+instance MonadState s m => MonadState s (ReaderT r m) where
+  get = lift get
+  put = lift . put
+  state = lift . state
+```
+
+This is very mechanical and boilerplaty. For these common transformers, these
+instances have all been written. However, every time you add an additional
+transformer - you need to write all these `n` instances where `n` is the number
+of interfaces you intend to use. Thus the `n^2` complexity.
+
+## Control structures for monad transformers
+For most of our monads that we create ourselves, they simply require this very
+mechanical boilerplate. This behavior looks like it could be captured by a
+typeclass (or two).
+
+One of our engineers, [Moisés](https://twitter.com/1akrmn), who previously
```

Comment:
When I joined Strats in Jan 2017, the codebase was already making heavy use of type classes for individual effects, e.g. `MonadTime`, `MonadDelay`, `MonadLog`, etc. but there was no solution to the n^2 problem. Monad transformers were providing instances for all the effects, relying on deriving to avoid as much boilerplate as possible. Alexis [article](https://lexi-lambda.github.io/blog/2017/04/28/lifts-for-free-making-mtl-typeclasses-derivable/) takes this approach to the extreme.

I made the point that introducing a new effect class required adding it to the deriving lists of all N transformers, which made engineers unwilling to add effects. and the approach could not scale. My solution to this was the passthrough instance, which requires a `MonadTransControl` transformer (or `MonadtTrans` for non-scoped effects). Since all `ReaderT` transformers are in `MonadTransControl` by definition unless the environment mentions the base monad, the codebase quickly gravitates towards `ReaderT` in order to avoid having to write instances manually.

Reply:
Thanks a lot for the insight! Added it to the post and linked to your comment 🙌